RLLib: ExternalEnv concurrent episode leaking issue

### What happened + What you expected to happen

I'm RLLib's external env with PPO to drive a physical system. For this reason, I am limited to one env with one worker. To ensure this, I build the algorithm (PPO) with `num_rollout_workers=1, batch_mode="complete_episodes", num_envs_per_worker=1` 

**The Bug**
And after a couple of episodes, I consistently get `AssertionError: Too many concurrent episodes, were some leaked? This ExternalEnv was created with max_concurrent=2`

The full stack trace:
```
Traceback (most recent call last):
  File "/home/aydin/Documents/TREC-RLT/scripts/pendulum/pendulum_external_replication.py", line 70, in <module>
    main()
  File "/home/aydin/Documents/TREC-RLT/scripts/pendulum/pendulum_external_replication.py", line 67, in main
    agent.train()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/tune/trainable/trainable.py", line 367, in train
    raise skipped from exception_cause(skipped)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/tune/trainable/trainable.py", line 364, in train
    result = self.step()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 749, in step
    results, train_iter_ctx = self._run_one_training_iteration()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/algorithms/algorithm.py", line 2623, in _run_one_training_iteration
    results = self.training_step()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/algorithms/ppo/ppo.py", line 318, in training_step
    train_batch = synchronous_parallel_sample(
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/execution/rollout_ops.py", line 85, in synchronous_parallel_sample
    sample_batches = worker_set.foreach_worker(
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/worker_set.py", line 696, in foreach_worker
    handle_remote_call_result_errors(remote_results, self._ignore_worker_failures)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/worker_set.py", line 73, in handle_remote_call_result_errors
    raise r.get()
ray.exceptions.RayTaskError(AssertionError): ray::RolloutWorker.apply() (pid=195731, ip=192.168.2.49, repr=<ray.rllib.evaluation.rollout_worker.RolloutWorker object at 0x7f0288f98970>)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/utils/actor_manager.py", line 183, in apply
    raise e
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/utils/actor_manager.py", line 174, in apply
    return func(self, *args, **kwargs)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/execution/rollout_ops.py", line 86, in <lambda>
    lambda w: w.sample(), local_worker=False, healthy_only=True
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/rollout_worker.py", line 900, in sample
    batches = [self.input_reader.next()]
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/sampler.py", line 92, in next
    batches = [self.get_data()]
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/sampler.py", line 285, in get_data
    item = next(self._env_runner)
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/evaluation/sampler.py", line 660, in _env_runner
    unfiltered_obs, rewards, dones, infos, off_policy_actions = base_env.poll()
  File "/home/aydin/anaconda3/envs/trec-rlt/lib/python3.9/site-packages/ray/rllib/env/external_env.py", line 372, in poll
    assert len(results[0]) <= limit, (
AssertionError: Too many concurrent episodes, were some leaked? This ExternalEnv was created with max_concurrent=1
```

(Note than max_concurrent is set to 2, because RLLib's polling algorithm checks whether the number of concurrent episodes is `<` limit, rather than `<=` limit.

### Versions / Dependencies

OS = Ubuntu 20.04
Python = 3.9
Ray = 2.2.0
Full output of `conda list` is [here](https://github.com/ray-project/ray/files/10510898/versions.txt)


### Reproduction script

Here is the env class and the driver script. Place the env in a file `pendulum_env_replication`, and run the driver script from the same directory. 

### Driver Script
```

from ray.tune import register_env
from ray.rllib.algorithms.ppo import PPOConfig
from ray.rllib.env.external_env import ExternalEnv
from pendulum_env_replication import PendulumExternalEnv


def main():
        
    register_env("pendulum-v0", lambda c: PendulumExternalEnv(c))
    agent = (
        PPOConfig()
        .resources(num_gpus=1)
        .rollouts(num_rollout_workers=1, batch_mode="complete_episodes", num_envs_per_worker=1)
        .environment(env="pendulum-v0")
        .build()
    )       

    while True:
        agent.train()

if __name__ == "__main__":
    main()
```
### Env
```
from ray.rllib.env.external_env import ExternalEnv
import time
import numpy as np
import gym

class PendulumExternalEnv(ExternalEnv):
    
    
    def __init__(self, config):
        
        # action space [pwm]
        current_limit: float = 1
        self.action_space = gym.spaces.Box(low=-current_limit, high=current_limit, shape=(1,), dtype=np.float32)
        
        # observation space [current angle (rad), target angle (rad), current angular velocity (rad/s)]
        self.angle_limit: float = 0.2
        observation_limit = np.array([self.angle_limit, self.angle_limit, 2 * np.pi])
        self.observation_space = gym.spaces.Box(low=-observation_limit, high=observation_limit, dtype=np.float32)
        self.observation = np.zeros(3, dtype=np.float32)
        
        ExternalEnv.__init__(self, self.action_space, self.observation_space, max_concurrent = 1)
                
        # environment variables
        self.episode_duration: float = 5
        self.episode_num: int = 0
        self.start_time = time.time()

        
    def run(self):
        while True:
            episode_id: str = self.start_episode()
            
            while not self._check_done():
                print("loopin")
                
                action = self.get_action(episode_id, self.observation)[0]
                self.log_returns(episode_id, 0)
            
            print("resetting")
            self.start_time = time.time()
            self.end_episode(episode_id, [0,0,0]) #FIXME: add actual obs to end_episode 
            
            
    def _check_done(self):
        if time.time() - self.start_time > self.episode_duration:
            return True
        
        if abs(self.observation[1]) > self.angle_limit:
            return True
        
        return False

```

### Issue Severity

High: It blocks me from completing my task.

--- a/rllib/env/external_env.py
+++ b/rllib/env/external_env.py
@@ -13,7 +13,6 @@ from ray.rllib.utils.typing import (
     EnvType,
     MultiEnvDict,
 )
-from ray.rllib.utils.deprecation import deprecation_warning
 
 if TYPE_CHECKING:
     from ray.rllib.models.preprocessors import Preprocessor
@@ -54,7 +53,6 @@ class ExternalEnv(threading.Thread):
         self,
         action_space: gym.Space,
         observation_space: gym.Space,
-        max_concurrent: int = None,
     ):
         """Initializes an ExternalEnv instance.
 
@@ -71,7 +69,6 @@ class ExternalEnv(threading.Thread):
         self._episodes = {}
         self._finished = set()
         self._results_avail_condition = threading.Condition()
-        self._max_concurrent_episodes = max_concurrent
 
     @PublicAPI
     def run(self):
@@ -376,11 +373,6 @@ class ExternalEnvWrapper(BaseEnv):
                 results = self._poll()
                 if not self.external_env.is_alive():
                     raise Exception("Serving thread has stopped.")
-        limit = self.external_env._max_concurrent_episodes
-        assert len(results[0]) < limit, (
-            "Too many concurrent episodes, were some leaked? This "
-            "ExternalEnv was created with max_concurrent={}".format(limit)
-        )
         return results
 
     @override(BaseEnv)

--- a/rllib/env/external_env.py
+++ b/rllib/env/external_env.py
@@ -376,6 +376,11 @@ class ExternalEnvWrapper(BaseEnv):
                 results = self._poll()
                 if not self.external_env.is_alive():
                     raise Exception("Serving thread has stopped.")
+        limit = self.external_env._max_concurrent_episodes
+        assert len(results[0]) < limit, (
+            "Too many concurrent episodes, were some leaked? This "
+            "ExternalEnv was created with max_concurrent={}".format(limit)
+        )
         return results
 
     @override(BaseEnv)

--- a/rllib/env/external_env.py
+++ b/rllib/env/external_env.py
@@ -13,6 +13,7 @@ from ray.rllib.utils.typing import (
     EnvType,
     MultiEnvDict,
 )
+from ray.rllib.utils.deprecation import deprecation_warning
 
 if TYPE_CHECKING:
     from ray.rllib.models.preprocessors import Preprocessor
@@ -53,15 +54,13 @@ class ExternalEnv(threading.Thread):
         self,
         action_space: gym.Space,
         observation_space: gym.Space,
-        max_concurrent: int = 100,
+        max_concurrent: int = None,
     ):
         """Initializes an ExternalEnv instance.
 
         Args:
             action_space: Action space of the env.
             observation_space: Observation space of the env.
-            max_concurrent: Max number of active episodes to allow at
-                once. Exceeding this limit raises an error.
         """
 
         threading.Thread.__init__(self)
@@ -377,11 +376,6 @@ class ExternalEnvWrapper(BaseEnv):
                 results = self._poll()
                 if not self.external_env.is_alive():
                     raise Exception("Serving thread has stopped.")
-        limit = self.external_env._max_concurrent_episodes
-        assert len(results[0]) <= limit, (
-            "Too many concurrent episodes, were some leaked? This "
-            "ExternalEnv was created with max_concurrent={}".format(limit)
-        )
         return results
 
     @override(BaseEnv)

--- a/rllib/env/external_env.py
+++ b/rllib/env/external_env.py
@@ -378,7 +378,7 @@ class ExternalEnvWrapper(BaseEnv):
                 if not self.external_env.is_alive():
                     raise Exception("Serving thread has stopped.")
         limit = self.external_env._max_concurrent_episodes
-        assert len(results[0]) < limit, (
+        assert len(results[0]) <= limit, (
             "Too many concurrent episodes, were some leaked? This "
             "ExternalEnv was created with max_concurrent={}".format(limit)
         )

--- a/rllib/env/external_env.py
+++ b/rllib/env/external_env.py
@@ -13,6 +13,7 @@ from ray.rllib.utils.typing import (
     EnvType,
     MultiEnvDict,
 )
+from ray.rllib.utils.deprecation import deprecation_warning
 
 if TYPE_CHECKING:
     from ray.rllib.models.preprocessors import Preprocessor
@@ -53,6 +54,7 @@ class ExternalEnv(threading.Thread):
         self,
         action_space: gym.Space,
         observation_space: gym.Space,
+        max_concurrent: int = None,
     ):
         """Initializes an ExternalEnv instance.
 
@@ -69,6 +71,14 @@ class ExternalEnv(threading.Thread):
         self._episodes = {}
         self._finished = set()
         self._results_avail_condition = threading.Condition()
+        if max_concurrent is not None:
+            deprecation_warning(
+                "The `max_concurrent` argument has been deprecated. Please configure"
+                "the number of episodes using the `rollout_fragment_length` and"
+                "`batch_mode` arguments. Please raise an issue on the Ray Github if "
+                "these arguments do not support your expected use case for ExternalEnv",
+                error=True,
+            )
 
     @PublicAPI
     def run(self):

--- a/rllib/env/external_multi_agent_env.py
+++ b/rllib/env/external_multi_agent_env.py
@@ -16,17 +16,14 @@ class ExternalMultiAgentEnv(ExternalEnv):
         self,
         action_space: gym.Space,
         observation_space: gym.Space,
-        max_concurrent: int = 100,
     ):
         """Initializes an ExternalMultiAgentEnv instance.
 
         Args:
             action_space: Action space of the env.
             observation_space: Observation space of the env.
-            max_concurrent: Max number of active episodes to allow at
-                once. Exceeding this limit raises an error.
         """
-        ExternalEnv.__init__(self, action_space, observation_space, max_concurrent)
+        ExternalEnv.__init__(self, action_space, observation_space)
 
         # We require to know all agents' spaces.
         if isinstance(self.action_space, dict) or isinstance(

--- a/rllib/env/tests/test_external_env.py
+++ b/rllib/env/tests/test_external_env.py
@@ -17,32 +17,27 @@ from ray.rllib.utils.test_utils import framework_iterator
 from ray.tune.registry import register_env
 
 
-def make_simple_serving(multiagent, superclass):
-    class SimpleServing(superclass):
-        def __init__(self, env):
-            superclass.__init__(self, env.action_space, env.observation_space)
-            self.env = env
+class SimpleServing(ExternalEnv):
+    def __init__(self, env, max_concurrent=100):
+        ExternalEnv.__init__(
+            self,
+            env.action_space,
+            env.observation_space,
+            max_concurrent=max_concurrent,
+        )
+        self.env = env
 
-        def run(self):
-            eid = self.start_episode()
-            obs, info = self.env.reset()
-            while True:
-                action = self.get_action(eid, obs)
-                obs, reward, terminated, truncated, info = self.env.step(action)
-                if multiagent:
-                    self.log_returns(eid, reward)
-                else:
-                    self.log_returns(eid, reward, info=info)
-                if terminated or truncated:
-                    self.end_episode(eid, obs)
-                    obs, info = self.env.reset()
-                    eid = self.start_episode()
-
-    return SimpleServing
-
-
-# generate & register SimpleServing class
-SimpleServing = make_simple_serving(False, ExternalEnv)
+    def run(self):
+        eid = self.start_episode()
+        obs, info = self.env.reset()
+        while True:
+            action = self.get_action(eid, obs)
+            obs, reward, terminated, truncated, info = self.env.step(action)
+            self.log_returns(eid, reward, info=info)
+            if terminated or truncated:
+                self.end_episode(eid, obs)
+                obs, info = self.env.reset()
+                eid = self.start_episode()
 
 
 class PartOffPolicyServing(ExternalEnv):

--- a/rllib/env/tests/test_external_env.py
+++ b/rllib/env/tests/test_external_env.py
@@ -18,12 +18,11 @@ from ray.tune.registry import register_env
 
 
 class SimpleServing(ExternalEnv):
-    def __init__(self, env, max_concurrent=100):
+    def __init__(self, env):
         ExternalEnv.__init__(
             self,
             env.action_space,
             env.observation_space,
-            max_concurrent=max_concurrent,
         )
         self.env = env
 
@@ -122,21 +121,6 @@ class TestExternalEnv(unittest.TestCase):
     def test_external_env_complete_episodes(self):
         ev = RolloutWorker(
             env_creator=lambda _: SimpleServing(MockEnv(25)),
-            default_policy_class=MockPolicy,
-            config=AlgorithmConfig().rollouts(
-                rollout_fragment_length=40,
-                batch_mode="complete_episodes",
-                num_rollout_workers=0,
-                enable_connectors=False,
-            ),
-        )
-        for _ in range(3):
-            batch = ev.sample()
-            self.assertEqual(batch.count, 50)
-
-    def test_external_env_max_concurrent(self):
-        ev = RolloutWorker(
-            env_creator=lambda _: SimpleServing(MockEnv(25), max_concurrent=1),
             default_policy_class=MockPolicy,
             config=AlgorithmConfig().rollouts(
                 rollout_fragment_length=40,

--- a/rllib/env/tests/test_external_env.py
+++ b/rllib/env/tests/test_external_env.py
@@ -134,6 +134,21 @@ class TestExternalEnv(unittest.TestCase):
             batch = ev.sample()
             self.assertEqual(batch.count, 50)
 
+    def test_external_env_max_concurrent(self):
+        ev = RolloutWorker(
+            env_creator=lambda _: SimpleServing(MockEnv(25), max_concurrent=1),
+            default_policy_class=MockPolicy,
+            config=AlgorithmConfig().rollouts(
+                rollout_fragment_length=40,
+                batch_mode="complete_episodes",
+                num_rollout_workers=0,
+                enable_connectors=False,
+            ),
+        )
+        for _ in range(3):
+            batch = ev.sample()
+            self.assertEqual(batch.count, 50)
+
     def test_external_env_truncate_episodes(self):
         ev = RolloutWorker(
             env_creator=lambda _: SimpleServing(MockEnv(25)),

--- a/rllib/env/tests/test_external_multi_agent_env.py
+++ b/rllib/env/tests/test_external_multi_agent_env.py
@@ -4,13 +4,28 @@ import unittest
 import ray
 from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
 from ray.rllib.env.external_multi_agent_env import ExternalMultiAgentEnv
-from ray.rllib.env.tests.test_external_env import make_simple_serving
 from ray.rllib.evaluation.rollout_worker import RolloutWorker
 from ray.rllib.evaluation.tests.test_rollout_worker import MockPolicy
 from ray.rllib.examples.env.multi_agent import BasicMultiAgent
 from ray.rllib.policy.sample_batch import SampleBatch
 
-SimpleMultiServing = make_simple_serving(True, ExternalMultiAgentEnv)
+
+class SimpleMultiServing(ExternalMultiAgentEnv):
+    def __init__(self, env):
+        ExternalMultiAgentEnv.__init__(self, env.action_space, env.observation_space)
+        self.env = env
+
+    def run(self):
+        eid = self.start_episode()
+        obs, info = self.env.reset()
+        while True:
+            action = self.get_action(eid, obs)
+            obs, reward, terminated, truncated, info = self.env.step(action)
+            self.log_returns(eid, reward)
+            if terminated or truncated:
+                self.end_episode(eid, obs)
+                obs, info = self.env.reset()
+                eid = self.start_episode()
 
 
 class TestExternalMultiAgentEnv(unittest.TestCase):

